lib/multi-writable.js

```diff
@@ -59,43 +123,256 @@ class MultiWriteable extends stream.Writable {
   }
 
   _write(chunk, enc, next) {
-    const str = chunk.toString();
-    this.buffer += str;
+    this.next = next;
+    this.buffer = Buffer.concat([this.buffer, chunk]);
+    this.handleNewBuffer();
+  }
+
+  handleNewBuffer() {
+    switch (this.state) {
+      case STATES.INSERT:
+        this.continueInsert();
+        break;
+      case STATES.CREATE:
+        this.runReadSchema();
+        break;
+      default:
+        this.determineState();
+        break;
+    }
+  }
+
+
+  // Handle No State
+
+  determineState() {
+    const indexOfCreate = this.buffer.indexOf(BUFFER_VALUES.CREATE_TABLE, this.bufferPosition);
+    const indexOfInsert = this.buffer.indexOf(BUFFER_VALUES.INSERT_INTO, this.bufferPosition);
+
+    if (indexOfCreate > -1 && (indexOfInsert === -1 || indexOfCreate < indexOfInsert)) {
+      this.state = STATES.CREATE;
+      this.runReadSchema();
+    } else if (indexOfInsert > -1 && (indexOfCreate === -1 || indexOfInsert < indexOfCreate)) {
+      this.state = STATES.INSERT;
+      this.runInsert();
+    } else {
+      this.state = null;
+      this.next();
+    }
+  }
 
-    const queue = this.buffer.split(this.delimeter);
-    const last = queue.pop();
 
-    this.buffer = last;
-    this.queue = this.queue.concat(queue);
-    this.flushQueueToAvailableWorkers(next);
+  // Handle Insert State
+
+  setTable(tableName) {
+    if (this.tables[tableName]) {
+      this.currentTable = tableName;
+    } else {
+      throw new Error(`The schema for table '${tableName}' has not been set yet. Your dump needs a CREATE TABLE query.`);
+    }
   }
 
-  flushQueueToAvailableWorkers(next) {
-    const nextWorkers = this.readyWorkers.splice(0, this.queue.length);
+  runInsert() {
+    const { INSERT_INTO, VALUES } = BUFFER_VALUES;
+
+    const indexOfInsert = this.buffer.indexOf(INSERT_INTO, this.bufferPosition);
+    const indexOfValues = this.buffer.indexOf(VALUES, indexOfInsert);
+
+    if (indexOfValues === -1) {
+      this.next();
+    } else {
+      const tablePos = indexOfInsert + INSERT_INTO.length;
+      const tableName = cleanVarName(this.buffer.slice(tablePos, indexOfValues).toString());
+      this.setTable(tableName);
+
+      this.bufferPosition = indexOfValues + VALUES.length;
+      this.continueInsert();
+    }
+  }
+
+  continueInsert() {
+    const { COMMAND_EXEC } = BUFFER_VALUES;
     const writes = [];
-    let queueItem;
     let row;
+    let worker;
+
+    let queueItem = this.getNextCommandParenSet();
+    while (queueItem && queueItem !== COMMAND_EXEC && this.readyWorkers.length) {
+      worker = this.readyWorkers.pop();
+      row = `[${queueItem}]`;
```

Comment:
This is coercing a buffer into a string. May be more performant to find a way to keep everything as a buffer, but perhaps for another PR